Much of the remaining library wording. (#176)
brevzin/cpp_proposals: 2996_reflection/reflection.md

```
:::

### `object_of`, `value_of`

::: std
```c++
namespace std::meta {
  consteval auto object_of(info r) -> info;
  consteval auto value_of(info r) -> info;
}
```
:::

If `r` is a reflection of a variable _usable in constant expressions_, then `object_of(r)` is a reflection of the object designated by the variable. For all other inputs, `object_of(r)` is not a constant expression.

::: std
```c++
int x;
int &y = x;

static_assert(^x != ^y);

> static_assert(object_of(^x) == object_of(^y));

Reviewer: So I think this description isn't quite right because `x` isn't "usable in constant expressions", but `y` is?

But I do think we want these two assertions to work, yeah.

Author: Good call - What I want is "variable designating an object having static storage duration".

consteval bool is_volatile(info r);
```

> [#]{.pnum} *Returns*: `true` if `r` designates a const or volatile type (respectively), a const- or volatile-qualified member function type (respectively), or an object, variable, or function with such a type. Otherwise, `false`.

Reviewer: ```cpp
struct C { int const x; };
```
Does `is_const(^C::x)` evaluate to `true` or `false`?

Author: Yeah, let's make that work.

Author: Also "member function types" aren't a thing, so we should just say "const- or volatile-qualified function type" (which ought to capture abominable functions).

[#]{.pnum} *Returns*: A reflection of the type of that entity.  If every declaration of that entity was declared with the same type alias (but not a template parameter substituted by a type alias), the reflection returned is for that alias.  Otherwise, if some declaration of that entity was declared with an alias it is unspecified whether the reflection returned is for that alias or for the type underlying that alias. Otherwise, the reflection returned shall not be a type alias reflection.

```cpp
consteval info object_of(info r);
```

> [#]{.pnum} *Mandates*: `r` is a reflection designating either an object or a variable usable in constante xpressions ([expr.const]).

Reviewer: Typo
```suggestion
[#]{.pnum} *Mandates*: `r` is a reflection designating either an object or a variable usable in constant expressions ([expr.const]).
```

Author: Done.

::: std
::: addu
```cpp
consteval size_t offset_of(info r);
```

> [#]{.pnum} *Mandates*: `r` is a reflection designating a non-static data member or base class.

Reviewer: Non-virtual base class maybe?

Author: Good call, done.

> [#]{.pnum} *Returns*: The offset in bytes, from the beginning of any object of type `parent_of(r)`, to the subobject associated with the entity reflected by `r`.

Reviewer: Drop the commas
```suggestion
[#]{.pnum} *Returns*: The offset in bytes from the beginning of any object of type `parent_of(r)` to the subobject associated with the entity reflected by `r`.
```

Author: Done.

```cpp
consteval size_t size_of(info r);
```

[#]{.pnum} *Mandates*: `r` is a reflection of a type, non-static data member, base class, object, or variable.

[#]{.pnum} *Returns* If `r` designates a type `T`, then `sizeof(T)`. Otherwise, `size_of(type_of(r))`.

```cpp
consteval size_t alignment_of(info r);
```

[#]{.pnum} *Mandates*: `r` is a reflection designating an object, variable, type, non-static data member, or base class.

[#]{.pnum} *Returns*: If `r` designates a type, object, or variable, then the alignment requirement of the entity. Otherwise, if `r` designates a base class, then `alignment_of(type_of(r))`. Otherwise, the alignment requirement of the subobject associated with the reflected non-static data member within any object of type `parent_of(r)`.

```cpp
consteval size_t bit_offset_of(info r);
```

[#]{.pnum} *Mandates*: `r` is a reflection designating a non-static data member.

> [#]{.pnum} *Returns*: The offset in bits, from the beginning of any object of type `parent_of(r)`, to the subobject associated with the entity reflected by `r`, minus the value of `offset_of(r);`

Reviewer: I'd rather write this as:

```suggestion
[#]{.pnum} Let `V` be the offset in bits from the beginning of an object of type `parent_of(r)` to the subobject associated with the entity reflected by `r`.

[#]{.pnum} *Returns*: `V - offset_of(r)`.
```

Also fixes a typo on any --> an.

I think this makes it more obvious that there's a subtraction going on, otherwise the "minus" is kind of far into the words.

Author: Done.

```cpp
consteval size_t bit_size_of(info r);
```

[#]{.pnum} *Mandates*: `r` is a reflection of a type, non-static data member, base class, object, or variable.

> [#]{.pnum} *Returns* If `r` designates a type, then the size in bits of any object having the reflected type. Otherwise, if `r` designates a variable, base class, or object, then `bit_size_of(type_of(r))`. Otherwise, the size in bits of the subobject associated with the reflected non-static data member within any object of type `parent_of(r)`.

Reviewer: Do we need to say something special about bit-fields?

Author: I hoped to capture that in the wording about "non-static data members". Just realized that the _width_ of a bit-field is a formal term, so let me state more explicitly in terms of that. Done.

:::
:::

### [meta.reflection.result] Expression result reflection {-}

::: std
::: addu
```cpp
template <typename T>
  consteval info reflect_value(T expr);
```

> [#]{.pnum} *Mandates*: `T` is a structural type. `T` is not a reference type. Any subobject of the value computed by `expr` having reference or pointer type designates an entity that is a permitted result of a constant expression.

Reviewer: Can combine these, and add a cross ref.
```suggestion
[#]{.pnum} *Mandates*: `T` is a structural type that is not a reference type. Any subobject of the value computed by `expr` having reference or pointer type designates an entity that is a permitted result of a constant expression ([expr.const]).
```

[#]{.pnum} *Returns*: A reflection of the value computed by an lvalue-to-rvalue conversion applied to `expr`. The type of the reflected value is the cv-unqualified version of `T`.

```cpp
template <typename T>
  consteval info reflect_object(T& expr);
```

[#]{.pnum} *Mandates*: `T` is not a function type. `expr` designates an entity that is a permitted result of a constant expression.

[#]{.pnum} *Returns*: A reflection of the object referenced by `expr`.

```cpp
template <typename T>
  consteval info reflect_function(T& expr);
```

[#]{.pnum} *Mandates*: `T` is a function type.

[#]{.pnum} *Returns*: `^fn`, where `fn` is the function referenced by `expr`.

```cpp
template <reflection_range R = span<info const>>
  consteval info reflect_invoke(info target, R&& args);
template <reflection_range R1 = span<info const>, reflection_range R2 = span<info const>>
  consteval info reflect_invoke(info target, R1&& tmpl_args, R2&& args);
```

> [#]{.pnum} Let `F` be the entity reflected by `target`, let `Arg0` be the entity reflected by the first element of `args` (if any), let `Args...` be the sequence of entities reflected by the elements of `args` excluding the first, and let `TArgs...` be the sequence of entities or aliases designated by the elements of `tmpl_args`.

Reviewer: "designated" by target?

Author: Keeping as is.

> [#]{.pnum} If `F` is a non-member function, a value of pointer to function type, or a value of closure type, then let `INVOKE-EXPR` be the expression `F(Arg0, Args...)`. Otherwise, if `F` is a member function or a value of pointer to member function type, then let `INVOKE-EXPR` be the expression `Arg0.F(Args...)` or `Arg0->F(Args...)` respectively. Otherwise, if `F` is a constructor for a class `C`, then let `INVOKE-EXPR` be the expression `C(Arg0, Args...)` for which only the constructor `F` is considered by overload resolution. Otherwise, if `F` is a non-member function template or a member function template, then let `INVOKE-EXPR` be the expression `F<TArgs...>(Arg0, Args...)` or `Arg0.template F<TArgs...>(Args...)` respectively. Otherwise, if `F` is a constructor template, then let `INVOKE-EXPR` be the expression `C(Arg0, Args...)` for which only the constructor `F` is considered by overload resolution, and `TArgs...` are inferred as explicit template arguments for `F`.

Reviewer: For the first part, I think can lean on `INVOKE(F, Arg0, Args...)` for the first part? Cause it could also be a pointer to member data and it'd be nice to just not have to invent new things. 

I think also that avoids needing to split the arguments, since you just get `INVOKE(F, Args...)` for the first overload and `F<TArgs...>(Args...)` for the second.

Author: Rebased onto INVOKE.
